chapter/methodology.md

```diff
@@ -0,0 +1,1116 @@
+Nickel's type checking implementation
+  ~ was adapted to pass AST nodes to the `Linearizer`.
+    Modifications to Nickel are minimal, comprising only few additional function calls and a slightly extended argument list.
+A stub implementation
+  ~ of the `Linearizer` trait is used during normal operation.
+    Since most methods of this implementation are `no-op`s, the compiler should be able to optimize away all `Linearizer` calls in release builds.
+
+<!-- TODO: caption -->
+```{.graphviz #fig:nls-nickel-structure caption="Interaction of Components"}
+digraph {
+  splines="ortho"
+  node [shape=record]
+
+
+
+  {
+    node[style=dashed]
+    nls [label="NLS"]
+    host [label=Analysis]
+  }
+
+  {
+    node[style=dotted]
+    nickel [label="Nickel"]
+    stub [label="Stub interface"]
+  }
+
+
+  als [label="Linearizer | \<T\>"]
+
+//   {rank=same; host; stub; }
+
+  
+  
+  host ->  nickel  [label="imports", constraint = false]
+  
+  stub ->  als [label="implements | T = ()"]
+  nickel ->  als  [label="calls"]
+  nickel ->  stub  [label="defines", style=dashed, color=grey]
+  nls ->  host  [label="defines", style=dashed, color=grey]
+  host ->  als [label="implements | T = Nickel"]
+
+}
+```
+
+
+#### Usage Graph
+
+At the core the linearization is a simple *linear* structure.
+Yet, it represents relationships of nodes on a structural level as a tree-like structure.
+Taking into account variable usage information adds back-edges to the original AST, yielding a graph structure.
+Both kinds of edges have to be encoded with the elements in the list.
+Alas, items have to be referred to using `id`s since the index of items cannot be relied on (such as in e.g., a binary heap), because the array is reordered to optimize access by source position.
+
+There are two groups of vertices in such a graph.
+**Declarations** are nodes that introduce an identifier, and can be referred to by a set of nodes.
+Referral is represented by **Usage** nodes.
+
+During the linearization process this graphical model is embedded into the items of the linearization.
+Hence, each `LinearizationItem` is associated with a kind representing the item's role in the graph (see: [@lst:nls-termkind-definition]).
+
+```{.rust #lst:nls-termkind-definition caption="Definition of a linearization items TermKind"}
+pub enum TermKind {
+    Declaration(Ident, Vec<ID>, ValueState),
+    Record(HashMap<Ident, ID>),
+    RecordField {
+        ident: Ident,
+        record: ID,
+        usages: Vec<ID>,
+        value: ValueState,
+    },
+
+    Usage(UsageState),
+
+    Structure,
+}
+
+pub enum UsageState {
+    Unbound,
+    Resolved(ID),
+    Deferred { parent: ID, child: Ident },
+}
+
+pub enum ValueState {
+    Unknown,
+    Known(ID),
+}
+```
+
+Variable bindings and function arguments
+  ~ are linearized using the `Declaration` variant which holds
+
+    - the bound identifier
+    - a list of `ID`s corresponding to its `Usage`s.
+    - its assigned value
```

Comment:
The bullet points here add more clarity than a long or multiple sentences I would argue.
Is it still preferable to write these out?